<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MLE assignment of (low-depth) samples to a tree — add_lowpass_sampled • MLLLP</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="MLE assignment of (low-depth) samples to a tree — add_lowpass_sampled" />
<meta property="og:description" content="MLE assignment of (low-depth) samples to a tree" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MLLLP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MLE assignment of (low-depth) samples to a tree</h1>
    
    <div class="hidden name"><code>add_lowpass_sampled.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>MLE assignment of (low-depth) samples to a tree</p>
    </div>

    <pre class="usage"><span class='fu'>add_lowpass_sampled</span><span class='op'>(</span>
  <span class='va'>tree</span>,
  <span class='va'>phydata</span>,
  <span class='va'>sample_data</span>,
  min_confidence <span class='op'>=</span> <span class='fl'>0</span>,
  vaf_bkgr <span class='op'>=</span> <span class='fl'>0.01</span>,
  purity_estimates <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>sample_data</span><span class='op'>)</span><span class='op'>)</span>,
  split_states <span class='op'>=</span> <span class='cn'>TRUE</span>,
  min_edge_length <span class='op'>=</span> <span class='fl'>0</span>,
  return_details <span class='op'>=</span> <span class='cn'>FALSE</span>,
  optimize_values <span class='op'>=</span> <span class='cn'>TRUE</span>,
  max_vaf_bkgr <span class='op'>=</span> <span class='fl'>0.1</span>,
  max_loss_frac <span class='op'>=</span> <span class='fl'>0</span>,
  loss_frac_init <span class='op'>=</span> <span class='fl'>0</span>,
  rescale_tree <span class='op'>=</span> <span class='cn'>TRUE</span>,
  nbtrials <span class='op'>=</span> <span class='fl'>20</span>,
  use_genetic_algorithm <span class='op'>=</span> <span class='cn'>FALSE</span>,
  control <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>tree</th>
      <td><p>A tree object (of class phylo).</p></td>
    </tr>
    <tr>
      <th>phydata</th>
      <td><p>Phylogenetic data used for the tree construction (of class phyDat). The attribute attr(*, "id") containng mutation ids has to be added (e.g., c("chr1:5_A/T", ...)</p></td>
    </tr>
    <tr>
      <th>sample_data</th>
      <td><p>A list of mutation data of samples to add the the tree. Each element must contain a data frame with the following columns: 'id' the mutation id as set in the 'phydata' argument, 'alt' (or 'alt_count') the number of mutated reads, 'depth' (or 'dp') the coverage of the site, 'cn_total' (or 'cn') the copy-number of the site, 'cn_mutated' (or 'mm') the number of mutated alleles (this is assumed to be 1 if missing).</p></td>
    </tr>
    <tr>
      <th>min_confidence</th>
      <td><p>(optional) minimum confidence in the the edge a sample is assigned to (default: 0).</p></td>
    </tr>
    <tr>
      <th>vaf_bkgr</th>
      <td><p>(optional) expected background mutation rate of unmutated sites (default 0.01).</p></td>
    </tr>
    <tr>
      <th>purity_estimates</th>
      <td><p>(optional) purity of the samples. Must be the same length as the sample_data list (default 1.0, pure samples).</p></td>
    </tr>
    <tr>
      <th>split_states</th>
      <td><p>(optional) logical flag indicating if the position along the edge should be estimated. Otherwise samples are only assigned to nodes (default: true).</p></td>
    </tr>
    <tr>
      <th>min_edge_length</th>
      <td><p>(optional) numeric value indicating the length of the edges to the added tips relative to the tree height (default: 0.01).</p></td>
    </tr>
    <tr>
      <th>return_details</th>
      <td><p>(optional) logical flag indicating if detailed information should be returned (default: false).</p></td>
    </tr>
    <tr>
      <th>optimize_values</th>
      <td><p>(optional) logical flag indicating if purity, background mutation rate, (and loss fraction) should be optimized (default: true).</p></td>
    </tr>
    <tr>
      <th>max_vaf_bkgr</th>
      <td><p>(optional) maximum background mutation rate (default: 0.1).</p></td>
    </tr>
    <tr>
      <th>max_loss_frac</th>
      <td><p>(optional) maximum loss fraction (default: 0).</p></td>
    </tr>
    <tr>
      <th>loss_frac_init</th>
      <td><p>(optional) initial value of loss fraction (default: 0).</p></td>
    </tr>
    <tr>
      <th>rescale_tree</th>
      <td><p>(optional) flag indicating if the tree should be rescaled (default: true).</p></td>
    </tr>
    <tr>
      <th>nbtrials</th>
      <td><p>(optional) number of restarts for optimisation (default: 20).</p></td>
    </tr>
    <tr>
      <th>use_genetic_algorithm</th>
      <td><p>(optional) logical should DEoptim::DEoptim be used? (default: FALSE).</p></td>
    </tr>
    <tr>
      <th>control</th>
      <td><p>(optional) control passed to optimizer (default: NULL).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>unused arguments.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A tree object with the samples in 'sample_data' added to it.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


