<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>MLE assignment of (low-depth) samples to a tree — add_lowpass_sampled • MLLPT</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="MLE assignment of (low-depth) samples to a tree — add_lowpass_sampled"><meta property="og:description" content="MLE assignment of (low-depth) samples to a tree"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MLLPT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MLE assignment of (low-depth) samples to a tree</h1>
    
    <div class="hidden name"><code>add_lowpass_sampled.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>MLE assignment of (low-depth) samples to a tree</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">add_lowpass_sampled</span><span class="op">(</span>
  <span class="va">tree</span>,
  <span class="va">phydata</span>,
  <span class="va">sample_data</span>,
  min_confidence <span class="op">=</span> <span class="fl">0</span>,
  vaf_bkgr <span class="op">=</span> <span class="fl">0.01</span>,
  purity_estimates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.75</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sample_data</span><span class="op">)</span><span class="op">)</span>,
  min_edge_length <span class="op">=</span> <span class="fl">0</span>,
  return_details <span class="op">=</span> <span class="cn">TRUE</span>,
  optimize_values <span class="op">=</span> <span class="cn">TRUE</span>,
  max_vaf_bkgr <span class="op">=</span> <span class="fl">0.1</span>,
  max_loss_frac <span class="op">=</span> <span class="fl">0</span>,
  loss_frac_init <span class="op">=</span> <span class="fl">0</span>,
  rescale_tree <span class="op">=</span> <span class="cn">TRUE</span>,
  control <span class="op">=</span> <span class="cn">NULL</span>,
  n_bootstraps <span class="op">=</span> <span class="fl">0</span>,
  n_cores <span class="op">=</span> <span class="fl">1</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>tree</dt>
<dd><p>A tree object (of class phylo).</p></dd>
<dt>phydata</dt>
<dd><p>Phylogenetic data used for the tree construction (of class phyDat).</p>
<p>Note: The attribute attr(*, "id") containng mutation ids has to be added (e.g., c("chr1:5_A/T", ...)</p></dd>
<dt>sample_data</dt>
<dd><p>A list of mutation data of samples to add the the tree.</p>
<p>Each element must contain a data frame with the following columns:</p><ul><li><p>'id': The mutation id as set in the 'phydata' argument.</p></li>
<li><p>'alt' (or 'alt_count'): The number of mutated reads.</p></li>
<li><p>'depth' (or 'dp'): The coverage of the site.</p></li>
<li><p>'cn_total' (or 'cn'): The copy-number of the site.</p></li>
<li><p>'cn_mutated' (or 'mm'): The number of mutated alleles (this is assumed to be 1 if missing).</p></li>
</ul></dd>
<dt>min_confidence</dt>
<dd><p>(optional) minimum confidence in the the edge a sample is assigned to (default: 0).</p></dd>
<dt>vaf_bkgr</dt>
<dd><p>(optional) expected background mutation rate of unmutated sites (default 0.01).</p></dd>
<dt>purity_estimates</dt>
<dd><p>(optional) purity of the samples.
Must be the same length as the sample_data list (default 1.0, pure samples).</p></dd>
<dt>min_edge_length</dt>
<dd><p>(optional) numeric value indicating the length of the edges
to the added tips relative to the tree height (default: 0.01).</p></dd>
<dt>return_details</dt>
<dd><p>(optional) logical flag indicating if detailed
information should be returned (default: false).</p></dd>
<dt>optimize_values</dt>
<dd><p>(optional) logical flag indicating if purity, background mutation rate,
(and loss fraction) should be optimized (default: true).</p></dd>
<dt>max_vaf_bkgr</dt>
<dd><p>(optional) maximum background mutation rate (default: 0.1).</p></dd>
<dt>max_loss_frac</dt>
<dd><p>(optional) maximum loss fraction (default: 0).</p></dd>
<dt>loss_frac_init</dt>
<dd><p>(optional) initial value of loss fraction (default: 0).</p></dd>
<dt>rescale_tree</dt>
<dd><p>(optional) flag indicating if the tree should be rescaled (default: true).</p></dd>
<dt>control</dt>
<dd><p>(optional) control passed to optimizer (default: NULL).</p></dd>
<dt>n_bootstraps</dt>
<dd><p>(optional) Number of bootstraps to perform (default: 0).</p></dd>
<dt>n_cores</dt>
<dd><p>(optional) Number of cores to use for bootstrapping (default: 0).</p></dd>
<dt>...</dt>
<dd><p>unused arguments.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A tree object with the samples in 'sample_data' added to it.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Timon Heide.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

